### ISSUE TITLE
The rollback of StatefulSet can't be done by increasing rollingUpdate.partition

### ISSUE THREAD

**Issue by jinseu** [kubernetes/kubernetes#91676]
**What happened**:

I have updated StatefulSet using `rollingUpdate strategy`, when `updateStrategy.rollingUpdate.partition==8` (`replicas==10`) I found new container image has bug. I would roll back the updated pod，set `updateStrategy.rollingUpdate.partition=10` and delete the updated pod(`kubectl delete pod`).Then statefulset created pod failure, and reported warning event. Warning message is

```
create Pod voltest-8 in StatefulSet voltest failed error: Pod "voltest-8" is invalid: [spec.volumes[3].hostPath: Forbidden: may not specify more than 1 volume type, spec.containers[0].volumeMounts[3].name: Not found: "qs-databackup"]
```

**What you expected to happen**:

I expect StatefulSet can successfully create pod after being rolled back.

**How to reproduce it (as minimally and precisely as possible)**:

Step 1. Create a StatefulSet，of which the PodSpec has a hostPath volume named `host-etc`，then mount `host-etc` to container `voltest` path `/data/etc`

```
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  labels:
    k8s-app: voltest
    qcloud-app: voltest
  name: voltest
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 10
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: voltest
      qcloud-app: voltest
  serviceName: ""
  template:
    metadata:
      annotations:
        tke.cloud.tencent.com/networks: tke-route-eni
        tke.cloud.tencent.com/vpc-ip-claim-delete-policy: Never
      creationTimestamp: null
      labels:
        k8s-app: voltest
        qcloud-app: voltest
    spec:
      containers:
      - env:
        - name: Hello
          value: World20200526
        image: nginx:latest
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        name: mmtkeyarddockerb
        ports:
        - containerPort: 80
          protocol: TCP
        readinessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        securityContext:
          capabilities:
            add:
            - SYS_PTRACE
            - SYS_RESOURCE
          privileged: false
          procMount: Default
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /home/qspace/log
          name: qs-log
        - mountPath: /data/etc
          name: host-etc
        - mountPath: /data/backup
          name: qs-databackup
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: qcloudregistrykey
      - name: tencenthubkey
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 5
      volumes:
      - emptyDir: {}
        name: tmp
      - emptyDir: {}
        name: qs-log
      - hostPath:
          path: /etc
          type: Directory
        name: host-etc
      - emptyDir: {}
        name: qs-databackup
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
```

Step 2. Update StatefulSet

1. Add a new hostPath volume named `host-bin`，then mount `host-bin` to container `voltest` path `/data/bin`, 
2. Set statefulset `StatefulSetUpdateStrategy` as a rollingUpdate, value of `rollingUpdate.partition` is 8

```
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  labels:
    k8s-app: voltest
    qcloud-app: voltest
  name: voltest
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 10
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: voltest
      qcloud-app: voltest
  serviceName: ""
  template:
    metadata:
      annotations:
        tke.cloud.tencent.com/networks: tke-route-eni
        tke.cloud.tencent.com/vpc-ip-claim-delete-policy: Never
      creationTimestamp: null
      labels:
        k8s-app: voltest
        qcloud-app: voltest
    spec:
      containers:
      - env:
        - name: Hello
          value: World20200526
        image: nginx:latest
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        name: mmtkeyarddockerb
        ports:
        - containerPort: 80
          protocol: TCP
        readinessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        securityContext:
          capabilities:
            add:
            - SYS_PTRACE
            - SYS_RESOURCE
          privileged: false
          procMount: Default
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /home/qspace/log
          name: qs-log
        - mountPath: /data/etc
          name: host-etc
        - mountPath: /data/backup
          name: qs-databackup
        - mountPath: /data/bin
          name: host-bin
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: qcloudregistrykey
      - name: tencenthubkey
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 5
      volumes:
      - emptyDir: {}
        name: tmp
      - emptyDir: {}
        name: qs-log
      - hostPath:
          path: /etc
          type: Directory
        name: host-etc
      - hostPath:
          path: /bin
          type: Directory
        name: host-bin
      - emptyDir: {}
        name: qs-databackup
  updateStrategy:
    rollingUpdate:
      partition: 8
    type: RollingUpdate
```

Step 3. Update `rollingUpdate.partition: 10`，and then run `kubectl delete pod voltest-8`

```
updateStrategy:
    rollingUpdate:
      partition: 10
    type: RollingUpdate
```

**Anything else we need to know?**:

I reproduce it in my test environment，and think the func `pkg/controller/statefulset/stateful_set_utils.ApplyRevision` has a bug. 

Statefulset controller got current set by `currentSet, err := ApplyRevision(set, currentRevision)`，and the code is

```
func ApplyRevision(set *apps.StatefulSet, revision *apps.ControllerRevision) (*apps.StatefulSet, error) {
	clone := set.DeepCopy()
	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
	if err != nil {
		return nil, err
	}
	restoredSet := &apps.StatefulSet{}
	err = json.Unmarshal(patched, restoredSet)
	if err != nil {
		return nil, err
	}
	return restoredSet, nil
}
```
in my condition the revision.Data.Raw volumes
```
[{"emptyDir":{},"name":"tmp"},{"emptyDir":{},"name":"qs-log"},{"hostPath":{"path":"/etc","type":"Directory"},"name":"host-etc"},{"emptyDir":{},"name":"qs-databackup"}]
```
set.Spec.Template.Spec.Volumes
```
[{"name":"tmp","emptyDir":{}},{"name":"qs-log","emptyDir":{}},{"name":"host-etc","hostPath":{"path":"/etc","type":"Directory"}},{"name":"host-bin","hostPath":{"path":"/bin","type":"Directory"}},{"name":"qs-databackup","emptyDir":{}}]
```
clone.Spec.Template.Spec.Volumes
```
ApplyRevision_result [{"name":"tmp","emptyDir":{}},{"name":"qs-log","emptyDir":{}},{"name":"host-etc","hostPath":{"path":"/etc","type":"Directory"}},{"name":"qs-databackup","hostPath":{"path":"/bin","type":"Directory"},"emptyDir":{}}]
```

obviously in `clone.Spec.Template.Spec.Volumes`, item `qs-databackup` has specify more than 1 volume type, due to `strategicpatch.StrategicMergePatch` has a bug 

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.3-tk8s.3", GitCommit:"3b916a420d630c81eb65b20dc33b967593fc6603", GitTreeState:"clean", BuildDate:"2020-02-12T08:19:12Z", GoVersion:"go1.12.14", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Tencent Kubernetes Engine
- OS (e.g: `cat /etc/os-release`): Tencent tlinux
- Kernel (e.g. `uname -a`): 4.14.105-1-tlinux3-0011

/sig api-machinery
/sig apps

**Comment by fedebongio** [comment]
/cc @janetkuo 
Would you be able to take a look? Thanks!

**Comment by janetkuo** [comment]
cc @kow3ns

**Comment by jinseu** [comment]
@kow3ns help ~~~~~~~

**Comment by MikeSpreitzer** [comment]
Commit 3b916a420d630c81eb65b20dc33b967593fc6603 is not in github.com/kubernetes/kubernetes and I do not know where to find it.

**Comment by MikeSpreitzer** [comment]
I see different code in release 1.14.3: https://github.com/kubernetes/kubernetes/blob/v1.14.3/pkg/controller/statefulset/stateful_set_utils.go#L335-L346

**Comment by MikeSpreitzer** [comment]
I have reproduced the problem on release 1.14.3

**Comment by MikeSpreitzer** [comment]
I also tested on release 1.14.10 and the problem seems to be fixed.

I did a `git checkout v1.14.10` and the `hack/local-up-cluster.sh`.  Then I saw this:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.10", GitCommit:"575467a0eaf3ca1f20eb86215b3bde40a5ae617a", GitTreeState:"clean", BuildDate:"2020-07-28T07:22:28Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.10", GitCommit:"575467a0eaf3ca1f20eb86215b3bde40a5ae617a", GitTreeState:"clean", BuildDate:"2020-07-28T07:22:28Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"linux/amd64"}
```

Next I created the initial revision of the StatefulSet:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl create -f -
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  labels:
    k8s-app: voltest
    qcloud-app: voltest
  name: voltest
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 10
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: voltest
      qcloud-app: voltest
  serviceName: ""
  template:
    metadata:
      annotations:
        tke.cloud.tencent.com/networks: tke-route-eni
        tke.cloud.tencent.com/vpc-ip-claim-delete-policy: Never
      creationTimestamp: null
      labels:
        k8s-app: voltest
        qcloud-app: voltest
    spec:
      containers:
      - env:
        - name: Hello
          value: World20200526
        image: nginx:latest
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        name: mmtkeyarddockerb
        ports:
        - containerPort: 80
          protocol: TCP
        readinessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        securityContext:
          capabilities:
            add:
            - SYS_PTRACE
            - SYS_RESOURCE
          privileged: false
          procMount: Default
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /home/qspace/log
          name: qs-log
        - mountPath: /data/etc
          name: host-etc
        - mountPath: /data/backup
          name: qs-databackup
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: qcloudregistrykey
      - name: tencenthubkey
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 5
      volumes:
      - emptyDir: {}
        name: tmp
      - emptyDir: {}
        name: qs-log
      - hostPath:
          path: /etc
          type: Directory
        name: host-etc
      - emptyDir: {}
        name: qs-databackup
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
statefulset.apps/voltest created
```

waited until it was stable:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ kubectl get StatefulSet
NAME      READY   AGE
voltest   10/10   19s
```

Then made the change:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl replace -f -
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  labels:
    k8s-app: voltest
    qcloud-app: voltest
  name: voltest
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 10
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: voltest
      qcloud-app: voltest
  serviceName: ""
  template:
    metadata:
      annotations:
        tke.cloud.tencent.com/networks: tke-route-eni
        tke.cloud.tencent.com/vpc-ip-claim-delete-policy: Never
      creationTimestamp: null
      labels:
        k8s-app: voltest
        qcloud-app: voltest
    spec:
      containers:
      - env:
        - name: Hello
          value: World20200526
        image: nginx:latest
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        name: mmtkeyarddockerb
        ports:
        - containerPort: 80
          protocol: TCP
        readinessProbe:
          failureThreshold: 60
          periodSeconds: 1
          successThreshold: 1
          tcpSocket:
            port: 80
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        securityContext:
          capabilities:
            add:
            - SYS_PTRACE
            - SYS_RESOURCE
          privileged: false
          procMount: Default
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /home/qspace/log
          name: qs-log
        - mountPath: /data/etc
          name: host-etc
        - mountPath: /data/backup
          name: qs-databackup
        - mountPath: /data/bin
          name: host-bin
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: qcloudregistrykey
      - name: tencenthubkey
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 5
      volumes:
      - emptyDir: {}
        name: tmp
      - emptyDir: {}
        name: qs-log
      - hostPath:
          path: /etc
          type: Directory
        name: host-etc
      - hostPath:
          path: /bin
          type: Directory
        name: host-bin
      - emptyDir: {}
        name: qs-databackup
  updateStrategy:
    rollingUpdate:
      partition: 8
    type: RollingUpdate
statefulset.apps/voltest replaced
```

Waited until it was stable:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl get StatefulSet
NAME      READY   AGE
voltest   10/10   55s
```

Then tweaked the partition then deleted the pod.
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl patch StatefulSet voltest -p '{"spec": {"updateStrategy": {"rollingUpdate": {"partition": 10} } } }'
statefulset.apps/voltest patched

mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl get StatefulSet
NAME      READY   AGE
voltest   10/10   75s

mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl delete Pod voltest-8
pod "voltest-8" deleted

mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl get StatefulSet
NAME      READY   AGE
voltest   10/10   95s
```

Observed replacement pod in place:
```
mspreitz@mjd-dev-1back:~/go/src/k8s.io/kubernetes/_output/bin$ ./kubectl get Pod
NAME        READY   STATUS    RESTARTS   AGE
voltest-0   1/1     Running   0          2m24s
voltest-1   1/1     Running   0          2m24s
voltest-2   1/1     Running   0          2m24s
voltest-3   1/1     Running   0          2m24s
voltest-4   1/1     Running   0          2m24s
voltest-5   1/1     Running   0          2m24s
voltest-6   1/1     Running   0          2m24s
voltest-7   1/1     Running   0          2m24s
voltest-8   1/1     Running   0          58s
voltest-9   1/1     Running   0          104s
```

**Comment by behos** [comment]
We were looking at this today and realized that the root cause was in fact https://github.com/kubernetes/kubernetes/issues/78420

Which was fixed in 1.16.

Maybe you are hitting the same issue as well?
